python/ray/tests/test_gcs_ha_e2e.py

```diff
@@ -0,0 +1,190 @@
+                f"Container {self.name} has already exited before "
+                "we noticed it was ready",
+            )
+
+        if self.status != "running":
+            return False
+
+        networks = self._container.attrs["NetworkSettings"]["Networks"]
+        for (_, n) in networks.items():
+            if not n["IPAddress"]:
+                return False
+
+        if "Ray runtime started" in super().logs():
+            return True
+        return False
+
+    def client(self):
+        port = self.ports["8000/tcp"][0]
+        return HTTPConnection(f"localhost:{port}")
+
+
+gcs_network = network(driver="bridge")
+
+redis_image = fetch(repository="redis:latest")
+
+redis = container(
+    image="{redis_image.id}",
+    network="{gcs_network.name}",
+    command=(
+        "redis-server --save 60 1 --loglevel" " warning --requirepass 5241590000000000"
+    ),
+)
+
+header_node = container(
+    image="ray_ci:v1",
+    name="gcs",
+    network="{gcs_network.name}",
+    command=["ray", "start", "--head", "--block", "--num-cpus", "0"],
+    environment={"RAY_REDIS_ADDRESS": "{redis.ips.primary}:6379"},
+    wrapper_class=Container,
+    ports={
+        "8000/tcp": None,
+    },
+)
+
+worker_node = container(
+    image="ray_ci:v1",
+    network="{gcs_network.name}",
+    command=["ray", "start", "--address", "gcs:6379", "--block"],
+    environment={"RAY_REDIS_ADDRESS": "{redis.ips.primary}:6379"},
+    wrapper_class=Container,
+    ports={
+        "8000/tcp": None,
+    },
+)
+
+
+@pytest.fixture
+def docker_cluster(header_node, worker_node):
+    yield (header_node, worker_node)
+
+
+scripts = """
+import ray
+import json
+from fastapi import FastAPI
+app = FastAPI()
+from ray import serve
+ray.init(address="auto", namespace="g")
+
+@serve.deployment(name="Counter", route_prefix="/api", version="v1")
+@serve.ingress(app)
+class Counter:
+    def __init__(self):
+        self.count = 0
+
+    @app.get("/")
+    def get(self):
+        return {{"count": self.count}}
+
+    @app.get("/incr")
+    def incr(self):
+        self.count += 1
+        return {{"count": self.count}}
+
+    @app.get("/decr")
+    def decr(self):
+        self.count -= 1
+        return {{"count": self.count}}
+
+    @app.get("/pid")
+    def pid(self):
+        import os
+        return {{"pid": os.getpid()}}
+
+serve.start(detached=True, dedicated_cpu=True)
+
+Counter.options(num_replicas={num_replicas}).deploy()
+"""
+
+check_script = """
+import requests
+import json
+if {num_replicas} == 1:
+    b = json.loads(requests.get("http://127.0.0.1:8000/api/").text)["count"]
+    for i in range(5):
+        response = requests.get("http://127.0.0.1:8000/api/incr")
+        assert json.loads(response.text) == {{"count": i + b + 1}}
+
+pids = {{
+    json.loads(requests.get("http://127.0.0.1:8000/api/pid").text)["pid"]
+    for _ in range(5)
+}}
+
+print(pids)
+assert len(pids) == {num_replicas}
+"""
+
+
+def test_ray_server_basic(docker_cluster):
```

Comment:
It seems we need to fix the failure on Linux, and skip the test on Windows or MacOS?

Reply:
Yeah, you are right.